src/libraries/Microsoft.PowerFx.Connectors/Internal/InjectServiceProviderFunction.cs

```diff
@@ -0,0 +1,50 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.PowerFx.Core.App.ErrorContainers;
+using Microsoft.PowerFx.Core.Functions;
+using Microsoft.PowerFx.Core.Localization;
+using Microsoft.PowerFx.Core.Types;
+using Microsoft.PowerFx.Core.Utils;
+using Microsoft.PowerFx.Syntax;
+using Microsoft.PowerFx.Types;
+
+namespace Microsoft.PowerFx.Connectors
+{
+    internal class InjectServiceProviderFunction : TexlFunction, IAsyncConnectorTexlFunction
+    {
+        public InjectServiceProviderFunction()
+            : base(DPath.Root, nameof(InjectServiceProviderFunction), nameof(InjectServiceProviderFunction), (loc) => nameof(InjectServiceProviderFunction), FunctionCategories.REST, DType.EmptyTable, 0, 1, 1)
+        {
+        }
+
+        public override bool IsSelfContained => true;
+
+        public override bool IsAsync => true;
+
+        public override IEnumerable<TexlStrings.StringGetter[]> GetSignatures()
+        {
+            yield return new[] { TexlStrings.TableArg1 };
+        }
+
+        public Task<FormulaValue> InvokeAsync(FormulaValue[] args, IServiceProvider context, CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            // Creation of the table with its context
+            ConnectorTableValueWithServiceProvider connectorTableWithServiceProvider = new ConnectorTableValueWithServiceProvider(args[0] as ConnectorTableValue, context);
+            return Task.FromResult<FormulaValue>(connectorTableWithServiceProvider);
+        }
```

Comment:
Could we remove ConnectorTableValueWithServiceProvider altogether and just do:
(This is clearer about lifespans, and would also ensure the cancellation token is passed, which it is not today)

        public async Task<FormulaValue> InvokeAsync(FormulaValue[] args, IServiceProvider context, CancellationToken cancellationToken)
        {
            cancellationToken.ThrowIfCancellationRequested();

            var runtimeContext = context.GetService<BaseRuntimeConnectorContext>();

            // Creation of the table with its context
            ConnectorTableValueWithServiceProvider connectorTableWithServiceProvider = new ConnectorTableValueWithServiceProvider(args[0] as ConnectorTableValue, context);

            ConnectorTableValue table = (ConnectorTableValue)args[0]; // IRRewriter should gaurantee this is always true
            var tabularService = table._tabularService;
            var rows = await tabularService.GetItemsAsync(runtimeContext, cancellationToken).ConfigureAwait(false);

            // Create in-memory 
            var inMemTable = new InMemoryTableValue(table.IRContext, rows);

            return inMemTable;
        }

Reply:
Current approach defers GetRows()…  will be useful for FirstN() …

Reply:
Ok -but then we need to double check the caching in ConnectorTableValueWithServiceProvider._cachedRows

Reply:
We only cache per expression

Reply:
beware, an expression may call multiple times. 

`First(Table, 5); First(Table, 10)`
is a single expression. 

Again - I think the code works today (because the cache will just always pull all rows), -  but we should be aware of this nuance because this would be easy to break.

Reply:
See SQL_Tabular test, lines 2000-2004